packages/react/src/components/Popover/index.js

```diff
@@ -23,20 +32,142 @@ const Popover = React.forwardRef(function Popover(props, ref) {
+    const conditions = {
+      left: rect.x < -16,
+      top: rect.y < -16,
+      right: rect.x + (rect.width - 16) > document.documentElement.clientWidth,
+      bottom:
+        rect.y + (rect.height - 16) > document.documentElement.clientHeight,
+    };
+
+    if (
+      !conditions.left &&
+      !conditions.top &&
+      !conditions.right &&
+      !conditions.bottom
+    ) {
+      setAutoAligned(false);
+      return;
+    }
+
+    const alignments = [
+      'top',
+      'top-left',
+      'right-bottom',
+      'right',
+      'right-top',
+      'bottom-left',
+      'bottom',
+      'bottom-right',
+      'left-top',
+      'left',
+      'left-bottom',
+      'top-right',
+    ];
+
+    // Creates the prioritized list of options depending on ideal alignment coming from `align`
+    const options = [align];
+    let option =
+      alignments[(alignments.indexOf(align) + 1) % alignments.length];
+
+    while (option) {
+      if (options.includes(option)) {
+        break;
+      }
+      options.push(option);
+      option = alignments[(alignments.indexOf(option) + 1) % alignments.length];
+    }
+
+    function isVisible(alignment) {
+      popover.current.classList.add(`${prefix}--popover--${alignment}`);
+
+      const rect = floating.current.getBoundingClientRect();
+
+      // Check if popover is not visible to the left of the screen
+      if (rect.x < -16) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      // Check if popover is not visible at the top of the screen
+      if (rect.y < -16) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      // Check if popover is not visible to right of screen
+      if (rect.x + (rect.width - 16) > document.documentElement.clientWidth) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      // Check if popover is not visible to bottom of screen
+      if (rect.y + (rect.height - 16) > document.documentElement.clientHeight) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+      return true;
+    }
+
+    let alignment = null;
+
+    for (let i = 0; i < options.length; i++) {
+      const option = options[i];
+
+      if (isVisible(option)) {
+        alignment = option;
+        break;
+      }
+      alignment = align;
+    }
+
+    if (alignment) {
+      setAutoAligned(true);
+      setAutoAlignment(alignment);
+    }
+  }, [autoAligned, align, autoAlign, prefix]);
+
   return (
-    <BaseComponent {...rest} className={className} ref={ref}>
-      {children}
-    </BaseComponent>
+    <PopoverContext.Provider value={{ floating }}>
```

Comment:
Since this is a context provider, it might help to either `useMemo` the object here to prevent unnecessary child re-renders or pass in `floating` directly since it will have a stable object identity (and prevent re-renders)